File: android/src/main/kotlin/com/cloudwebrtc/webrtc/proxy/RtpReceiverProxy.kt
package com.cloudwebrtc.webrtc.proxy

import org.webrtc.RtpReceiver

/**
 * Wrapper around an [RtpReceiver].
 *
 * @param receiver  Underlying [RtpReceiver].
 */
class RtpReceiverProxy(receiver: RtpReceiver) : Proxy<RtpReceiver> {
    /**
     * Actual underlying [RtpReceiver].
     */
    override var obj: RtpReceiver = receiver

    /**
     * [MediaStreamTrackProxy] of this [RtpReceiverProxy].
     */
    private var track: MediaStreamTrackProxy? = null

    init {
        syncWithObject()
    }

    override fun syncWithObject() {
        syncMediaStreamTrack()
    }

    /**
     * @return  Unique ID of the underlying [RtpReceiver].
     */
    fun id(): String {
        return obj.id()
    }

    /**
     * Notifies [RtpReceiverProxy], that it's [MediaStreamTrackProxy] is ended.
     */
    fun ended() {
        track?.observableEventBroadcaster()?.onEnded()
    }

    /**
     * Synchronizes the [MediaStreamTrackProxy] of this [RtpReceiverProxy] with
     * the underlying [RtpReceiver].
     */
    private fun syncMediaStreamTrack() {
        val newReceiverTrack = obj.track()
        if (newReceiverTrack == null) {
            track = null
        } else {
            if (track == null) {
                track = MediaStreamTrackProxy(newReceiverTrack)
            } else {
                track!!.replace(newReceiverTrack)
            }
        }
    }
}

File: ios/Classes/proxy/RtpReceiverProxy.swift
import WebRTC

/// Wrapper around an `RTCRtpReceiver`.
class RtpReceiverProxy {
  /// Actual underlying [RtpReceiver].
  private var receiver: RTCRtpReceiver

  /// `MediaStreamTrackProxy` of this `RtpReceiverProxy`.
  private var track: MediaStreamTrackProxy

  /// Creates a new `RtpReceiverProxy` with a provided `RTCRtpReceiver`.
  init(receiver: RTCRtpReceiver) {
    self.receiver = receiver
    self.track = MediaStreamTrackProxy(track: self.receiver.track!, deviceId: nil, source: nil)
  }

  /// Returns ID of this `RtpReceiverProxy`.
  func id() -> String {
    self.receiver.receiverId
  }

  /// Returns `MediaStreamTrackProxy` of this `RtpReceiverProxy`.
  func getTrack() -> MediaStreamTrackProxy {
    self.track
  }

  /// Notifies `RtpReceiverProxy` about its `MediaStreamTrackProxy` being removed from the receiver.
  func notifyRemoved() {
    self.track.notifyEnded()
  }
}
